File: solutions-in-cpp/B-Frog_2.cpp
#include<iostream>
#include<vector>

using namespace std;

#define ll long long 

const int INF = 1e9+5;

int main(){
	//freopen("tree.in","r",stdin);
    //freopen("tree.out","w",stdout);
    //freopen("input.txt", "r", stdin);
    //freopen("out.txt", "w", stdout);
    ios_base::sync_with_stdio(0);
    //cin.tie(0);
	
	int n, k;
	cin >> n >> k;
	vector<int> height(n);
	for(int i = 0; i < n; ++i){
		cin >> height[i];
	}
	vector<int> dp(n, INF);
	dp[0] = 0;
	for(int i = 0; i < n; ++i){
		for(int j = i+1; j <= i+k; ++j){
			if(j < n){
				dp[j] = min(dp[j], dp[i] + abs(height[i]-height[j]));
			}
		}
	}
	cout << dp[n-1] << endl;
}

File: solutions-in-python/B-Frog_2.py
INF = (10**9)+5

n, k = map(int, input().split())
height = []*n 

height = list(map(int, input().split()))

dp = [INF]*n # initializing dp with INF with size of n
dp[0] = 0

for i in range(n):
	for j in range(i+1, i+k+1):
		if j < n:
			dp[j] = min(dp[j], dp[i] + abs(height[i]-height[j]))

print(dp[n-1])			
